launcher.py

```diff
@@ -0,0 +1,206 @@
+    # exposed by dstack's guest agent). To avoid installing another Python depdendency, namely
+    # requests-unixsocket, we just use curl.
+    extend_rtmr3_json = '{"event": "launcher-image-digest","payload": "%s"}' % image_digest.split(':')[1]
+    proc = run(['curl', '--unix-socket', '/var/run/dstack.sock', '-X', 'POST',
+                'http://dstack/EmitEvent', '-H', 'Content-Type: application/json', '-d',
+                extend_rtmr3_json])
+
+    if proc.returncode:
+        raise RuntimeError("extending rtmr3 failed with error code %d" % proc.returncode)
+
+    # Get quote after extending RTMR3 with the image digest
+    proc = run(['curl', '--unix-socket', '/var/run/dstack.sock', '-X', 'POST',
+                'http://dstack/GetQuote', '-H', 'Content-Type: application/json', '-d',
+                '{"report_data": ""}'], capture_output=True)
+
+    if proc.returncode:
+        raise RuntimeError("getting quote failed with error code %d" % proc.returncode)
+
+    logging.info("Quote: %s" % proc.stdout.decode('utf-8').strip())
+
+    # Build the docker command we use to start the app, i.e., mpc node
+    docker_cmd = ['docker', 'run']
+
+    for env_file in ['/tapp/.host-shared/.user-config']:
+        if os.path.isfile(env_file):
+            docker_cmd += ['--env-file', env_file]
+
+    # Forward ports from container to CVM
+    # This port is the MPC node port. It must be forwarded and cannot be remapped.
+    # TODO Adjust and/or make configurable according to `port_override` from the mpc node.
+    docker_cmd += ['-p', '11780:11780']
+
+    docker_cmd += ['-v', '/tapp:/tapp:ro']
+    docker_cmd += ['-v', '/var/run/dstack.sock:/var/run/dstack.sock']
+    docker_cmd += ['-v', 'shared-volume:/mnt/shared']
+    docker_cmd += ['-v', 'mpc-data:/data']
+
+    # TODO Remove after testing.
+    docker_cmd += ['--add-host', 'mpc-node-0.service.mpc.consul:66.220.6.113']
+    docker_cmd += ['--add-host', 'mpc-node-1.service.mpc.consul:57.129.144.117']
+
+    # Image digest must be last.
+    docker_cmd += ['--detach', image_digest]
+
+    logging.info("docker cmd %s", " ".join(docker_cmd))
+
+    # Start the app.
+    proc = run(docker_cmd)
+
+    if proc.returncode:
+        raise RuntimeError("docker run non-zero exit code %d", proc.returncode)
+
+def get_manifest_digest(registry_url: str, image_name: str, tags: list[str], image_digest: str):
+    '''Given an `image_digest` returns a manifest digest.
+
+       `docker pull` requires a manifest digest. This function translates an image digest into a manifest digest by talking to the Docker registry.
+
+       API doc for image registry https://distribution.github.io/distribution/spec/api/
+    '''
+    if not tags:
+        raise Exception(f"No tags found for image {image_name}")
+
+    # We need an authorization token to fetch manifests.
+    # TODO this still has the registry hard-coded in the url. also, if we use a different registry, we need a different auth-endpoint.
```

Comment:
Yes, we've agreed the single registry is good enough for now but we need to revisit this later.